### ISSUE TITLE
Support referencing the same embeddable type multiple times per table

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12608]
While embeddables can match different column names per table, each embeddable can be contained only once in each table. For example, with a table like this:

```sql
CREATE TABLE e (
  id INT NOT NULL,
  
  amount NUMERIC(20, 2) NOT NULL,
  currency CHAR(3) NOT NULL,
  x_amount NUMERIC(20, 2) NOT NULL,
  x_currency CHAR(3) NOT NULL,
  
  CONSTRAINT pk_e PRIMARY KEY (id)
);
```

It seems natural to have two references to a `MONETARY_AMOUNT` embeddable, with or without the `X_` prefix:

```xml
<embeddable>                                                                 
    <name>MONETARY_AMOUNT</name>                                             
    <fields>                                                                 
        <field><name>AMOUNT</name><expression>AMOUNT</expression></field>    
        <field><name>CURRENCY</name><expression>CURRENCY</expression></field>
    </fields>                                                                
</embeddable>                                                                
```

One would hope that something like this is possible (with better ergonomics, of course):

```xml
<embeddable>                                                                   
    <name>MONETARY_AMOUNT</name>                                               
    <referencingName>X_MONETARY_AMOUNT</referencingName>                       
    <fields>                                                                   
        <field><name>AMOUNT</name><expression>X_AMOUNT</expression></field>    
        <field><name>CURRENCY</name><expression>X_CURRENCY</expression></field>
    </fields>                                                                  
</embeddable>                                                                  
```

But this just produces a warning, and the duplicate declaration is ignored:

```
[WARNING] Embeddable configuration : Table PUBLIC.E already has embeddable <name>MONETARY_AMOUNT</name><referencingName>X_MONETARY_AMOUNT</referencingName><replacesFields>false</replacesFields><fields><field><name>AMOUNT</name><expression>X_AMOUNT</expression></field><field><name>CURRENCY</name><expression>X_CURRENCY</expression></field></fields>
```

The workaround is to copy the entire embeddable and have two incompatible classes, or to let the embeddables extend some common interface, at least.

----
See also:
- https://stackoverflow.com/q/69823707/521799

**Comment by alf** [comment]
The StackOverflow question is more about the ergonomics of such configuration. Is this also in the pipeline or is there some existing mechanism I can leverage?

**Comment by lukaseder** [comment]
I'm not sure what you mean, @alf?

**Comment by alf** [comment]
As you mention above:

> One would hope that something like this is possible (with better ergonomics, of course):
> …

I’m interested in the ergonomics here. We have the metadata in our db to generate the above configuration, but I’m not sure how to best go about doing that. Most of the examples in the documentation are based on static xml configuration. I haven’t seen any good examples of configuring the code generation based on more advanced queries/code against the database we’re generating against. Not sure if that makes things clearer.

**Comment by lukaseder** [comment]
I see, well, what I meant by this is that one would expect *some* way to configure reusing the same embeddable multiple times per table, and then I gave a simple example with a strawman XML syntax. A better syntax would be offered, once this feature is implemented. But it isn't implemented yet, so I don't think the ergonomics discussion is useful just yet.

> We have the metadata in our db to generate the above configuration, but I’m not sure how to best go about doing that. Most of the examples in the documentation are based on static xml configuration. I haven’t seen any good examples of configuring the code generation based on more advanced queries/code against the database we’re generating against. Not sure if that makes things clearer.

That sounds like it's unrelated to embeddable types? It's just XML. You can generate it using Java code, XSLT, etc., or you can use programmatic code generation configuration: https://www.jooq.org/doc/latest/manual/code-generation/codegen-programmatic/

There are many ways to achieve this. Not sure if this helps?

**Comment by alf** [comment]
Yes, it’s unrelated to the problem of not being able to use the same embeddable type twice in one table. The pointer to that page in the documentation is helpful, I can take it from there.

**Comment by lukaseder** [comment]
I get it now :) I didn't realise your Stack Overflow question was about something entirely different... I'll update the answer, too. Sorry for the confusion.

**Comment by alf** [comment]
I’m perpetually confused, so no harm done. Let me know if you need anything more from me regarding use case etc. ;)

**Comment by lukaseder** [comment]
Well, I'm always curious about use-cases! Are you using embeddables mainly for multi-column values or also for single-column values (to add more semantics to values)? What other things would improve your dev lives?

**Comment by alf** [comment]
In short, we’re working with a database that was modeled using the ORM/NIAM method 25 years ago. That model was then used to generate an Oracle database schema in 3NF, including row, table and database constraints. Over the past 25 years the database has been well cared for. Larger modifications has been done by going back to the original NIAM model, and then implemented true to style with the code generator used initially.

As a result, there is a lot of the original semantics of the NIAM model “hidden” in the naming conventions of the resulting database. As I’ve learned more about jOOQ, I believe we might be able to bring to light the original design that until now only is visible to those who really understand the system.

I hope this makes sense. I’m happy to expand on this, with the caveat that I’m still trying to get to grips with the original design.

**Comment by alf** [comment]
Single column values can be handled by using custom mappings and bindings, so we’re mostly interested in this for composite values.

**Comment by lukaseder** [comment]
> As I’ve learned more about jOOQ, I believe we might be able to bring to light the original design that until now only is visible to those who really understand the system.

Ah yes, that makes sense. You can encode the naming conventions in jOOQ code generation rules and make them available as types. That's a very good use-case for embeddable types.

**Comment by alf** [comment]
Is this hard to implement? Is it possible to attempt a solution and send pull request? Unless it’s already in progress of course 🤞

**Comment by lukaseder** [comment]
I'm not sure yet how tricky it is to implement this. There are 3 scenarios:

1. This can be pulled off purely with code generation configuration. This would be the ideal, of course
2. In addition to code generation, some additional runtime logic is required (see comment below)
3. The feature request exposes a fundamental design flaw of embeddables where the generated `EmbeddableRecord` has a hard reference to actual columns of a single table. That's already a bit weird today, when the embeddable is referenced from multiple tables, but it does work. But if the embeddable is referenced multiple times from the same table, it might not work. Would have to check.

**Comment about runtime logic**

Embeddables have quite a few refactorings on the roadmap. In short, the whole SQL/JSON, SQL/XML `MULTISET` emulation could also work quite nicely for embeddables and resolve quite a few open issues. I'm not sure yet if these planned refactorings prevent/postpone a prior implementation of this feature here, in the sense that this feature could be much easier to implement if the refactorings are already available.

If this can be done with code generation alone, then the above comment doesn't apply. Otherwise, a PR doesn't really make sense as you don't have access to the regression tests. A lot of things can break in this area...

**Ergonomics**

Let me play around with a possible implementation today. Maybe I can think of some configuration design.

**Comment by lukaseder** [comment]
So, when looking at your SO question here: https://stackoverflow.com/q/69823707/521799

You have:

```sql
create table t (
  amount int, 
  unit varchar2(4), 
  amount_pend int, 
  unit_pend varchar(4)
);
```

And you'd like to do something along the lines of:

```xml
<embeddables>
    <embeddable>
        <name>MONETARY_AMOUNT</name>
        <referencingName>AMOUNT_WITH_UNIT</referencingName>
        <tables>T/tables>
        <fields>
            <field><expression>AMOUNT</expression></field>
            <field><expression>UNIT</expression></field>
        </fields>
    </embeddable>
    <embeddable>
        <name>MONETARY_AMOUNT</name>
        <referencingName>AMOUNT_WITH_UNIT_PEND</referencingName>
        <tables>T/tables>
        <fields>
            <field><name>AMOUNT</name><expression>AMOUNT_PEND</expression></field>
            <field><name>UNIT</name><expression>UNIT_PEND</expression></field>
        </fields>
    </embeddable>
</embeddables>
```

The design rules would be:

- If multiple identical `<name/>` configurations are found, then
  - They will produce the same generated `EmbeddableRecord`
  - Their `<fields>` must produce the same `fields/field/name` for all references
  - All `<referencingName>` evaluations (including the default one, which is the same as `<name/>`) must be distinct *per table*

Thinking about this design, I think it might be possible, indeed, to implement only with code generation. A lot of integration tests will be required to be sure we cover all edge cases. I'll try this today.

**Comment by lukaseder** [comment]
Not strictly related, I've thought about various aspects of "synthetic DDL" before: https://github.com/jOOQ/jOOQ/issues/9526

Embeddables are just UDT emulations (`OBJECT` type in Oracle). In Oracle, you could also write:

```sql
CREATE TYPE monetary_amount AS OBJECT (amount int, varchar2(4));
CREATE TABLE t (amount monetary_amount, amount_pend monetary_amount);
```

So, in a case like this, it would be interesting if jOOQ supported some sort of synthetic UDT type:

```sql
CREATE SYNTHETIC TYPE monetary_amount AS (amount int, varchar2(4));
CREATE TABLE t (amount monetary_amount, amount_pend monetary_amount);
```

And the `DDLDatabase`, as well as other means of interpreting the above would translate it to either the Oracle version with an actual UDT, or the one you used, and auto-generate the necessary code generation configuration. But that's just food for thought, for now.

**Comment by lukaseder** [comment]
For starters, the configuration doesn't cause any problems, it's simply ignored:

```
11:53:48,819 [WARNING] Unused embeddable        : <name>MONETARY_AMOUNT</name><referencingName>MONETARY_AMOUNT_X</referencingName><replacesFields>false</replacesFields><fields><field><name>AMOUNT</name><expression>AMOUNT_X</expression></field><field><name>AMOUNT</name><expression>CURRENCY_X</expression></field></fields>
```

**Comment by alf** [comment]
Thank you, your comments align well with my intentions.

Just to add to the discussion around UDTs. We could have used UDTs in Oracle here. The problem is that this requires us to either:

1. Change the core schema and thus all users of such. There are many users and so this is not a realistic task as of now.
2. Create a facade in the database using views and introduce the UDTs there. This forces us to create instead of-triggers on the resulting views (as Oracle is not smart enough to handle write through for us when we do “fancy stuff” in a view). This makes for a lot of trigger writing and so this is not a tempting solution either.

The best way forward, as far as I can see, is embeddable records or some other solution in the Java stack.

**Comment by lukaseder** [comment]
Yeah, I thought about the few caveats of using UDTs in Oracle. Nevertheless, it's always good to keep in mind that an embeddable is just a client side UDT emulation. For jOOQ users, they should work in a very similar fashion.

**Comment by lukaseder** [comment]
It seems I have been too defensive. The first implementation draft has uncovered 1 bug in the generation of DAOs: https://github.com/jOOQ/jOOQ/issues/13028, where generated method names use the embeddable `name`, instead of the `referencingName`

Other than that, this seems to work! I'll push this draft right away, and you should have it available in the next 3.17.0-SNAPSHOT nightly build. Your early feedback would be greatly appreciated!

In the meantime, I'll think of more edge cases to test...

**Comment by lukaseder** [comment]
This was integrated in jOOQ 3.17.0. Apart from your reported issue #13040, I cannot think of any other edge cases, yet. When further issues are found, bug fixes can be implemented.

Thanks again for your time discussing this feature request.
